Add  com.github.PopoutApps.popout3d

PopoutApps opened the pull request:
### Please confirm your submission meets all the criteria

- [X ] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [X ] My pull request follows the instructions at [App Submission][submission].
- [ X] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [X ] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [X ] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [X ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [X] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

hfiguiere commented on org.flatpak.popout3d.yml:
```diff
@@ -0,0 +1,204 @@
+    buildsystem: meson
+    sources:
+      - type: git
+        url: https://github.com/popout-apps/popout3d
```
You should checkout a tag or a specific commit

popout-apps replied:
"exalm" wrote the manifest for me, for which I'm very grateful. I've not used Git before, so I'll have to learn how to do that.

popout-apps commented on org.flatpak.popout3d.yml:
```diff
@@ -0,0 +1,204 @@
+app-id: com.github.popout3-apps.popout3d
```
I have seen hyphens in in names like org.kde.WaylandDecoration.QGnomePlatform-decoration, so I assumed it was alright. Is it just the first name, the name of the Git domain that can't have a hyphen?
If so, I'll change it and everything else

hfiguiere replied:
It is allowed in the last component of the id. :-/

PopoutApps replied:
I'm changing it to PopoutApps as '_' isn't allowed in the GitHub domain/account name.

hfiguiere replied:
For the record, I never said to change the GitHub account. Just to replace `-` with `_` in the app id.

PopoutApps replied:
I thought it had to be the same, and even if not, I thought it would be best not to cause confusion by having different names. Hopefully changing it doesn't cause any problems.

popout-apps commented on org.flatpak.popout3d.yml:
```diff
@@ -0,0 +1,204 @@
+    sources:
+      - type: git
+        url: https://github.com/Exiv2/exiv2.git
+        commit: e7ffd83af29187190da7f98dcbca8a4d70e19582
```
See below.

popout-apps commented on org.flatpak.popout3d.yml:
```diff
@@ -0,0 +1,204 @@
+# There is now a Flatpak for hugin which might not have been available when 
+# this manifest was written. There may be something useful there.
+
+runtime-version: "3.30"
```
I'll change "**sdk: org.gnome.Sdk**" to "**sdk: org.Freedesktop.Sdk**" and I'll try again with 3.38.

PopoutApps replied:
(Leaving out all the hugin stuff for a quick test of the rest makes it quicker to build.)
Have changed to this:
```
runtime: org.freedesktop.Platform
runtime-version: '19.08'
sdk: org.freedesktop.Sdk
```
with this result:
```
import gi                                    <application fails here<
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk
```
Python Gtk tutorial says "We import the Gtk module. The require_version() method ensures the namespace gets loaded with the given version. The gi.repository is the Python module for PyGObject. PyGObject (Python GObject introspection) contains Python bindings and support for gobject, glib, gtk and other Gnome libraries."

barthalion commented on org.flatpak.popout3d.yml:
```diff
@@ -0,0 +1,204 @@
+  - --share=ipc
+  - --socket=x11
+  - --socket=wayland
+  - --filesystem=xdg-run/dconf
+  - --filesystem=~/.config/dconf:ro
+  - --talk-name=ca.desrt.dconf
+  - --env=DCONF_USER_CONFIG_DIR=.config/dconf
```
```suggestion
```

barthalion commented on org.flatpak.popout3d.yml:
```diff
@@ -0,0 +1,204 @@
+copy-icon: true
+finish-args:
+  - --share=ipc
+  - --socket=x11
```
```suggestion
  - --socket=fallback-x11
```

barthalion commented on org.flatpak.popout3d.yml:
```diff
@@ -0,0 +1,204 @@
+  - --filesystem=~/.config/dconf:ro
+  - --talk-name=ca.desrt.dconf
+  - --env=DCONF_USER_CONFIG_DIR=.config/dconf
+  - --filesystem=host:rw
```
You should use GtkFileChooserNative to open and save files.

PopoutApps replied:
Is that mandatory or a recommendation? It will take me a while to learn itt, so I'd prefer to get this sorted first and do that in the next version.

ChrisOfBristol replied:
> filechooserdialog = Gtk.**FileChooserNative**(title="Select a file", action=Gtk.FileChooserAction.OPEN)
![b](https://user-images.githubusercontent.com/20310498/104966132-13819700-59d8-11eb-8967-6cf7c9984802.png)
opens a completely new window with an extra icon to the list of open applications - very clumsy and confusing to use. It doesn't allow adding a reference to a parent window, so I tried
> filechooserdialog = Gtk.**FileChooserNative.new**('Select a file', **self.window**, Gtk.FileChooserAction.OPEN,'OK','Cancel')
('window' is the name of the application's top-level window)
no better, it still opens a completely separate window. 

Rather than use something that looks absolutely awful on any system including the native system, I have gone back to the original which looks great on its native system but might be less than perfectly matched on another system. 

Unless you have a suggestion for a better way to do this?

PopoutApps replied:
> filechooserdialog = Gtk.**FileChooserNative**(title="Select a file", action=Gtk.FileChooserAction.OPEN)

![b](https://user-images.githubusercontent.com/20310498/104966132-13819700-59d8-11eb-8967-6cf7c9984802.png)

opens a completely new window with an extra icon to the list of open applications - very clumsy and confusing to use. It doesn't allow adding a reference to a parent window, so I tried
> filechooserdialog = Gtk.**FileChooserNative.new**('Select a file', **self.window**, Gtk.FileChooserAction.OPEN,'OK','Cancel')

'window' is the name of the application's top-level window. This is no better, it still opens a completely separate window. 

Rather than use something that looks absolutely awful on any system including the native system, I will go back to the original which looks great on its native system but might be less than perfectly matched on another system. 

Unless you have a suggestion for a better way to do this?

nedrichards commented on org.flatpak.popout3d.yml:
```diff
@@ -0,0 +1,196 @@
+        sha256: d80ee30805f10bf898df9ce7495f46e88990e8f90b5c2b917722d87526252b2c
+
+  # Hugin deps  
+  - name: glu
```
you should pull in the shared-modules versions of glew and glu https://github.com/flathub/shared-modules

PopoutApps replied:
I've done that, but it fails, I don't know whether the failure is related to this change.
...
config.status: executing libtool commands
  CC       src/libutil/glue.lo
  CC       src/libutil/error.lo
  CC       src/libutil/mipmap.lo
src/libutil/error.c:60:10: error: ‘NULL’ undeclared here (not in a function)
   60 |    { ~0, NULL } /* end of list indicator */
      |          ^~~~
src/libutil/error.c:34:1: note: ‘NULL’ is defined in header ‘<stddef.h>’; did you forget to ‘#include <stddef.h>’?
   33 | #include <GL/glu.h>
  +++ |+#include <stddef.h>
   34 | 
make: *** [Makefile:1119: src/libutil/error.lo] Error 1
make: *** Waiting for unfinished jobs....
Error: module glu: Child process exited with code 2

nedrichards commented on org.flatpak.popout3d.yml:
```diff
@@ -0,0 +1,196 @@
+      cxxflags: "-std=c++0x"
+    sources:
+      - type: archive
+        url: https://github.com/wxWidgets/wxWidgets/releases/download/v3.0.4/wxWidgets-3.0.4.tar.bz2
```
does this work with the latest upstream 'stable' wxwidgets? (3.0.5.1)?

PopoutApps replied:
Changed, will try later.

nedrichards commented on org.flatpak.popout3d.yml:
```diff
@@ -0,0 +1,196 @@
+    buildsystem: meson
+    sources:
+      - type: git
+        url: https://github.com/PopoutApps/popout3d
```
this needs to refer to a specific commit, tag or release.

PopoutApps replied:
Now tagged.

hfiguiere commented:
The manifest file should be named `com.github.PopoutApps.popout3d.yml`

Or it won't work.

hfiguiere commented on org.flatpak.popout3d.yml:
```diff
@@ -0,0 +1,167 @@
+
+  # Hugin deps  
+  - shared-modules/glew/glew.json
+  - shared-modules/glu/glu-9.0.0.json
```
But you are missing the submodule.

PopoutApps replied:
Manifest renamed.
Could you explain what and from where please? Do you mean glew.json and glu-9.0.0.json? They are in the PopoutApps/popout3d repository. Should they be in the PopoutApps/flathub repository? Should they be in directories?

hfiguiere replied:
No.

See https://github.com/flathub/flathub/wiki/App-Requirements#shared-modules

hfiguiere commented on org.flatpak.popout3d.yml:
```diff
@@ -55,8 +55,8 @@ modules:
-        url: https://github.com/wxWidgets/wxWidgets/releases/download/v3.0.4/wxWidgets-3.0.4.tar.bz2
-        sha256: 96157f988d261b7368e5340afa1a0cad943768f35929c22841f62c25b17bf7f0
+        url: https://github.com/wxWidgets/wxWidgets/releases/download/v3.0.5.1/wxWidgets-3.0.5.1.tar.bz2
+        sha256: 406ac736f61d88a3a866aa501e01e408a642c6e7
```
This is not a sha256, it is too short.

(the build will fail)

Example:
https://github.com/flathub/net.sourceforge.Hugin/blob/master/net.sourceforge.Hugin.json#L54

PopoutApps replied:
That was what was on the webpage, I've replaced it with the one the error message said it was looking for.

hfiguiere replied:
They are sha1 not sha256

hfiguiere commented:
> The manifest file should be named `com.github.PopoutApps.popout3d.yml`
> 
> Or it won't work.

I'll reiterate, since you renamed it incorrectly.

PopoutApps commented:
> The manifest file should be named `com.github.PopoutApps.popout3d.yml`
> 
> Or it won't work.
Done.

hfiguiere commented:
bot, build com.github.PopoutApps.popout3d

hfiguiere commented on com.github.PopoutApps.popout3d.yml:
```diff
@@ -0,0 +1,173 @@
+app-id: com.github.PopoutApps.popout3d
+
+runtime: org.freedesktop.Sdk
```
This should be `org.freedesktop.Platform`

PopoutApps replied:
Done.

barthalion commented on com.github.PopoutApps.popout3d.yml:
```diff
@@ -0,0 +1,173 @@
+  - --share=ipc
+  - --socket=fallback-x11
+  - --socket=wayland
+  - --filesystem=host:rw
```
```suggestion
  - --filesystem=host
```

This is the default.

PopoutApps replied:
Done.

barthalion commented:
bot, build com.github.PopoutApps.popout3d

hfiguiere commented:
the build fail because the version of `glu` you build needs patches or needs to be updated (`shared-modules` has the right version and that's what it recommended to use)

And `glew` too.

bilelmoussaoui commented:
bot, build com.github.PopoutApps.popout3d

PopoutApps commented:
The failure appears to be in building align_image_stack (the last module before the program itself) relating to exiv2 (which is two modules back). Is there any useful information there which I could include in a query to exiv2 support?

```
[62/304] Building CXX object src/hugin_base/CMakeFiles/huginbase.dir/panodata/SrcPanoImage.cpp.o
FAILED: src/hugin_base/CMakeFiles/huginbase.dir/panodata/SrcPanoImage.cpp.o 
/usr/bin/c++ -DGLEW_STATIC -Dhuginbase_EXPORTS -Isrc -I../src/hugin_base -I../src/celeste -Isrc/celeste -I../src -I/app/include/OpenEXR -I../src/foreign -O2 -g -pipe -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -O2 -g -DNDEBUG -fPIC -fopenmp -pthread -std=gnu++11 -MD -MT src/hugin_base/CMakeFiles/huginbase.dir/panodata/SrcPanoImage.cpp.o -MF src/hugin_base/CMakeFiles/huginbase.dir/panodata/SrcPanoImage.cpp.o.d -o src/hugin_base/CMakeFiles/huginbase.dir/panodata/SrcPanoImage.cpp.o -c ../src/hugin_base/panodata/SrcPanoImage.cpp
../src/hugin_base/panodata/SrcPanoImage.cpp: In member function ‘bool HuginBase::SrcPanoImage::readEXIF()’:
../src/hugin_base/panodata/SrcPanoImage.cpp:358:30: error: expected unqualified-id before ‘&’ token
  358 |     catch (const Exiv2::Error& e)
      |                              ^
../src/hugin_base/panodata/SrcPanoImage.cpp:358:30: error: expected ‘)’ before ‘&’ token
  358 |     catch (const Exiv2::Error& e)
      |           ~                  ^
      |                              )
../src/hugin_base/panodata/SrcPanoImage.cpp:358:30: error: expected ‘{’ before ‘&’ token
../src/hugin_base/panodata/SrcPanoImage.cpp:358:32: error: ‘e’ was not declared in this scope
  358 |     catch (const Exiv2::Error& e)
      |                                ^
[63/304] Building CXX object src/hugin_base/CMakeFiles/huginbase.dir/panodata/Panorama.cpp.o
ninja: build stopped: subcommand failed.
Error: module align_image_stack: Child process exited with code 1
```

hfiguiere commented:
bot, build com.github.PopoutApps.popout3d

hfiguiere commented on com.github.PopoutApps.popout3d.yml:
```diff
@@ -0,0 +1,173 @@
+
+  # Hugin deps
+  - shared-modules/glew/glew.json
+  - shared-modules/glu/glu-9.0.0.json
```
Still the name doesn't match.... so it can't be found.

PopoutApps replied:
Fixed. Have also changed Exiv2 module "sources" and am getting somewhere with tests of that on my PC.

PopoutApps replied:
This has failed to build again. My bit for the Python stuff is okay, the problem lies with Hugin/align_image_stack. Any idea where I can get some help with this? I've posted a query on https://answers.launchpad.net/hugin/+question/695021 with no response so far.
I know this must be possible because there is a manifest here for Hugin
https://github.com/flathub/net.sourceforge.Hugin/blob/master/net.sourceforge.Hugin.json
and a Flatpak here
https://github.com/flathub/flathub/pull/1988

```
FAILED: src/hugin1/hugin/hugin 
..
..
view]+0x620): undefined reference to `wxGLCanvasX11::SwapBuffers()'
/usr/lib/gcc/x86_64-unknown-linux-gnu/10.2.0/../../../../x86_64-unknown-linux-gnu/bin/ld: src/hugin1/hugin/CMakeFiles/hugin.dir/GLViewer.cpp.o:(.data.rel.ro._ZTV10GLOverview[_ZTV10GLOverview]+0x638): undefined reference to `wxGLCanvas::GetXWindow() const'
/usr/lib/gcc/x86_64-unknown-linux-gnu/10.2.0/../../../../x86_64-unknown-linux-gnu/bin/ld: warning: creating DT_TEXTREL in a PIE
collect2: error: ld returned 1 exit status
[259/304] Building CXX object src/hugi...MakeFiles/PTBatcherGUI.dir/Batch.cpp.o
ninja: build stopped: subcommand failed.
Error: module align_image_stack: Child process exited with code 1

```

hfiguiere replied:
I'm the maintainer of the Hugin flathub :-)

as I said: more recent version of Hugin works. What you can try is to build wxWidget the exact same way as in the Hugin manifest, and if that doesn't work, using an older version known to work.

PopoutApps replied:
I have tried with all the URLs copied from the "sort of" manifest file on Launchpad. it doesn't work.

I appear to have come up against an unsolvable problem! When I build a Debian package with a dependency on Hugin, all I have to do is state that it "depends on" the Hugin Debian package. Flatpak does not allow me to state that it "depends” on the Hugin Flatpak. 

Flatpak requires me compile (I have written my app in an interpreted language and am not used to compilers) someone else's package (that I know nothing about) written in a language that I know nothing about (I have written my application in Python not C). I haven't got a snowball's chance in hell of achieving this.

I don't know which statements from the "sort of" manifest file on Launchpad would be valid and how to convert that format into a manifest.yml.

There might be a way - as you know Hubert, Hugin already has a Flatpak for which there must be a manifest.yml from which I might be able to copy the relevant bits. Could you give me a link to this please, otherwise this project is doomed...

hfiguiere commented:
bot, build com.github.PopoutApps.popout3d

hfiguiere commented:
Not sure why the link is failing. I build a more recent version of hugin without problem. Maybe wxWidget it too recent for this older version?

Also it should be noted that `align_image_stack` which is the binary you are interested in doesn't depend en wx, so it hugin could be build without wx, that would solve a lot more problems.

PopoutApps commented:
That makes sense to me since align_image_stack hasn't got a GUI. Building it now.

PopoutApps commented:
It's saying that it can't find wxwidgets. Can I tell it not to bother with it?
LOG
```
Determining if the function log1p exists passed with the following output:
Change Dir: /run/build/align_image_stack/_flatpak_build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/ninja cmTC_7deed && [1/2] Building C object CMakeFiles/cmTC_7deed.dir/CheckFunctionExists.c.o
<command-line>: warning: conflicting types for built-in function ‘log1p’; expected ‘double(double)’ [-Wbuiltin-declaration-mismatch]
/usr/share/cmake-3.19/Modules/CheckFunctionExists.c:7:3: note: in expansion of macro ‘CHECK_FUNCTION_EXISTS’
    7 |   CHECK_FUNCTION_EXISTS(void);
      |   ^~~~~~~~~~~~~~~~~~~~~
/usr/share/cmake-3.19/Modules/CheckFunctionExists.c:1:1: note: ‘log1p’ is declared in header ‘<math.h>’
  +++ |+#include <math.h>
    1 | #ifdef CHECK_FUNCTION_EXISTS
[2/2] Linking C executable cmTC_7deed
```
OUTPUT
```
========================================================================
Building module align_image_stack in /home/chris/.flatpak-builder/build/align_image_stack-1
========================================================================
-- The C compiler identification is GNU 10.2.0
-- The CXX compiler identification is GNU 10.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Current HG revision is 5abfb4de7961
-- Assuming this is a tarball (release) build for 2018.0.0
-- Looking for log1p
-- Looking for log1p - found
CMake Error at /usr/share/cmake-3.19/Modules/FindPackageHandleStandardArgs.cmake:218 (message):
  Could NOT find wxWidgets (missing: wxWidgets_LIBRARIES
  wxWidgets_INCLUDE_DIRS)
Call Stack (most recent call first):
  /usr/share/cmake-3.19/Modules/FindPackageHandleStandardArgs.cmake:577 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.19/Modules/FindwxWidgets.cmake:1016 (find_package_handle_standard_args)
  CMakeLists.txt:166 (FIND_PACKAGE)


-- Configuring incomplete, errors occurred!
See also "/run/build/align_image_stack/_flatpak_build/CMakeFiles/CMakeOutput.log".
Error: module align_image_stack: Child process exited with code 1

```

PopoutApps commented:
I've updated the manifest. I'd be interested to see whether you can make any conclusions from the result.
My next step will be to look at https://github.com/flathub/net.sourceforge.Hugin/blob/master/net.sourceforge.Hugin.json
{
    "app-id": "net.sourceforge.Hugin",
    "runtime": "org.freedesktop.Platform",
    "runtime-version": "20.08",
    "sdk": "org.freedesktop.Sdk",
    "command": "hugin",
..
..
take as much as possible/relevant and convert the format for mine.

barthalion commented:
bot, build com.github.PopoutApps.popout3d

hfiguiere commented:
bot, build com.github.PopoutApps.popout3d

PopoutApps commented:
Have abandoned using align-image-stack. This version has to manage without it.

barthalion commented:
bot, build com.github.PopoutApps.popout3d

barthalion commented on popout3d_metainfo.xml:
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
You need to reference this file in your manifest and install it at `/app/share/metainfo/${FLATPAK_ID}.metainfo.xml`.

barthalion replied:
Also your metainfo file should include [OARS](https://hughsie.github.io/oars/) and `<releases>` section.

PopoutApps replied:
Reference - todo. Install it - done.
OARS - done, <releases> done.

barthalion commented on shared-modules/glew/glew.json:
```diff
@@ -0,0 +1,30 @@
+{
```
You should add `shared-modules` as git submodule with `git submodule add https://github.com/flathub/shared-modules.git`.

PopoutApps replied:
Studying how to do this...

PopoutApps commented:
Please do not tell the bot to build this yet, as I have found a problem I am still trying to solve.

barthalion commented on com.github.PopoutApps.popout3d.yml:
```diff
@@ -0,0 +1,36 @@
+
+command: popout3d
+rename-desktop-file: popout3d.desktop
+rename-icon: popout3d
```
```suggestion
rename-icon: popout3d_logo
```

PopoutApps replied:
Logo - I was getting in a right mess with this. This used to work, but I removed the _logo from everything as I thought it was non-standard, I can't understand where it's coming from now. Commited as you suggested.

Have to stop work on this now.

barthalion replied:
Yeah, these are not errors, just a message that flatpak found some icon but it doesn't match the ID.

PopoutApps replied:
_...just a message that flatpak found some icon but it doesn't match the ID._
Didn't understand this, but:
_Yeah, these are not errors,_
This is what matters.

barthalion commented:
At first glance the suggestion above should fix the build. That being said, the app doesn't start because `/app/bin/popout3d` is not marked as executable, and then when I fixed it, it fails again as follows:

```
[📦 com.github.PopoutApps.popout3d bin]$ python3 popout3d 
Traceback (most recent call last):
  File "popout3d", line 969, in <module>
    main = GUI()	
  File "popout3d", line 850, in __init__
    self.builder.add_from_file(self.glade)	   
gi.repository.GLib.Error: g-file-error-quark: Failed to open file “/home/chris/debian/popout3d/popout3d.glade”: No such file or directory (
```

PopoutApps commented:
Logo - I'm getting in a right mess with this. This used to work, but I removed the _logo from everything as I thought it was non-standard, I can't understand where it's coming from now.

I've recently changed it to suit someone who want to run it uninstalled, I added a variable which has to be set the beginning depending on whether it will be packaged or not. It's getting the wrong directory because you are running it in a terminal rather than as a flatpak. On line-6, if you set the variable to **package = none** it should work.

barthalion commented:
Apps can detect if they're running in a flatpak sandbox by checking if the `/.flatpak-info` file exists.

PopoutApps commented:
> Apps can detect if they're running in a flatpak sandbox by checking if the `/.flatpak-info` file exists.
That's a much better way of doing it, I've changed the code and will build it again here when I have understood what's going on with the logo.

PopoutApps commented:
My problems with the _logo were caused by my not understanding tags. I've sorted those and the directories out. All that remains of what has been spotted so far is the executable not having the executable property. I'm wading through the Meson documentation to find out how to fix that.

nedrichards commented:
btw. you can use the bot commands as well once you're happy with it to generate a test build.

PopoutApps commented:
I didn't realise that.
Does it give you/the system more work - is it best avoided until I'm confident I'm getting somewhere with bug fixing? 
If not, how do I do it, is there docs somwhere?

barthalion commented:
> Does it give you/the system more work - is it best avoided until I'm confident I'm getting somewhere with bug fixing?

The build infrastructure is here to be used so while it's nice if you test your changes locally, it's not a problem if you also test everything here. There's no single tool to reproduce entire CI "test suite" locally yet.

bot, build com.github.PopoutApps.popout3d

barthalion commented:
I don't know much about meson but this seems suspicious: https://github.com/PopoutApps/popout3d/blob/008f26f759d17a1ba9c25adc4d51f45db7dbe667/meson.build#L44

The final file should be called `com.github.PopoutApps.popout3d.metainfo.xml`.

PopoutApps commented:
bot, build com.github.PopoutApps.popout3d

PopoutApps commented:
Can't make any sense of the error report.

nedrichards commented:
This error means that the Appstream data (the metadata that populates the website and software stores so people can discover your application) isn't being compiled correctly. This is usually because all the things (desktop file, appdata/metainfo file, icons) haven't been named after the app-id (com.github.PopoutApps.popout3d). I can take a quick peek, it's usually a fairly easy fix.

nedrichards commented:
Yep, your meson.build is putting the metainfo in the "wrong" place so it's not being detected on https://github.com/PopoutApps/popout3d/blob/main/meson.build#L42

@bilelmoussaoui has a helpful example here on how to do this correctly in meson (you can obviously ignore the rust specific bits) https://belmoussaoui.com/article/8-how-to-flatpak-a-rust-application

PopoutApps commented:
OK, what about this:
     - pip3 install --no-index --find-links="file://${PWD}" --prefix=${FLATPAK_DEST} Pillow Pillow-PIL
The $ makes me think of a Bash variable, should I have replaced this with something, or does it pick it up itself from somewhere?

PopoutApps commented:
bot, build com.github.PopoutApps.popout3d

PopoutApps commented:
bot, build com.github.PopoutApps.popout3d

PopoutApps commented:
bot, build com.github.PopoutApps.popout3d

PopoutApps commented:
bot, build com.github.PopoutApps.popout3d

PopoutApps commented:
There seems to me to be a contradiction here. The documentation says:
"AppData files should be named with the application ID and the .appdata.xml file extension, and should be placed in /app/share/metainfo/. For example:
/app/share/metainfo/org.gnome.Dictionary.appdata.xml"

So this is what I have in meson.build:
(data_dir is /app/share)
	'popout3d_metainfo.xml', rename : 'com.github.PopoutApps.popout3d.appdata.xml'
  install_dir: join_paths(datadir, 'metainfo') 
which should result in /app/share/metainfo/com.github.PopoutApps.popout3d.appdata.xml

But the build error report says:
stat: cannot stat 'builddir/*/share/app-info/xmls/com.github.PopoutApps.popout3d.xml.gz': No such file or directory

The directory it is looking for seems to be quite different. to what the docs say, but I'll put that down to the internal workings of the builder. However it seems to me it is looking for a different filename - without the 'appdata' mentioned in the documents. Should the 'appdata' be there or not?

hfiguiere commented:
What you describe you have in meson is not really what can be seen:

<img width="622" alt="image" src="https://user-images.githubusercontent.com/114441/106895877-cb908e80-66be-11eb-8b67-fd61f3793e98.png">


tl;dr it doesn't install the appdata file where it should be.

PopoutApps commented:
Thanks, my unfamiliarity with GitHub, forgot to do a new tag and release. Will do it now.

PopoutApps commented:
bot, build com.github.PopoutApps.popout3d

PopoutApps commented:
bot, build com.github.PopoutApps.popout3d

PopoutApps commented:
bot, build com.github.PopoutApps.popout3d

PopoutApps commented:
bot, build com.github.PopoutApps.popout3d

PopoutApps commented:
I'm  confused about names again. 
Error: _../meson.build:30:0: ERROR: File popout3d.desktop does not exist._
Where does the _popout3d.desktop_ come from?

desktop file has
_Name=Popout3D
Exec=com.github.PopoutApps.popout3d_

Manifest has:
_app-id: com.github.PopoutApps.popout3d
..
command: popout3d_

Meson.build has:
_configure_file(
  input: 'popout3d.in',
  output: 'popout3d',_

Should the manifest command and the meson.build output file both be _com.github.PopoutApps.popout3d_

PopoutApps commented:
bot, build com.github.PopoutApps.popout3d

PopoutApps commented:
bot, build com.github.PopoutApps.popout3d

PopoutApps commented:
I don't understand the error.

 argv: b'zgrep "<id>com.github.PopoutApps.popout3d\\(\\.\\w\\+\\)*\\(.desktop\\)\\?</id>" builddir/*/share/app-info/xmls/com.github.PopoutApps.popout3d.xml.gz'

nedrichards commented:
It's looking for this part of the appdata file https://github.com/flathub/flathub/wiki/AppData-Guidelines#id I'll take another look in a local compile.

nedrichards commented:
bot, build com.github.PopoutApps.popout3d

nedrichards commented:
OK, cool, this matches what I get in my local build when testing locally (as suggested on that page above)
> flatpak run org.freedesktop.appstream-glib validate app/files/share/appdata/com.github.PopoutApps.popout3d.appdata.xml

So, the screenshot - that github link won't download - you need to use the raw link from github: https://github.com/PopoutApps/popout3d/raw/main/popout3d.png but even then this "screenshot" will fail validation. Per the appstream docs https://www.freedesktop.org/software/appstream/docs/chap-Quickstart.html you need a 16:9 ratio file (ideally) that's at least 620px wide.

For the URL element you should show the homepage of your software, https://github.com/PopoutApps/popout3d would absolutely do for now.

Over and above this, now I have a local build I installed it and it's not launching with the following error on startup:
> mkdir: cannot create directory ‘/sysroot/home/nick/.popout3d/work/’: No such file or directory
Cannot create work directory

PopoutApps commented:
Thanks, I didn't know how to run that - so many things to learn at once. It passes with your recommended changes.
It's frustrating that it looks as though I never check anything - but when I install and run on my PC these things all run fine, PC seems not to be fussy enough.  Screenshot and URL added, Metadata file passes test, directory corrected.

bot, build com.github.PopoutApps.popout3d

PopoutApps commented:
Thanks Ned! 
It all works fine, except that the first time that I ran it, it wouldn't accept any input.

nedrichards commented:
/merge

nedrichards commented:
@PopoutApps no worries! We understand that it can be a bit of work. By and large flatpak itself deals with the transition from "works on my machine" to "works on everyone else's machine" - it's the extra app store metadata and information that lets people discover your app that needs some time. But then again, that's why we're here to help.

It's all good on my end so merging now. Have a quick look at the app maintenance process that the merge bot will tell you about but basically, pushes to master will automatically kick off new releases - you can check out the build progress here https://flathub.org/builds/ (and manually create or push a build if you want) once build is done it'll go into a temporary repo for a couple of hours for you to rest before it's automatically released. PRs on your repo will work like the PR here with test builds occuring automatically and going into temporary repos so you can experiment with changes without affecting users. If you've got any questions then it's quickest to pop onto #flatpak on Freenode IRC.

PopoutApps commented:
@hfiguiere hfiguiere, @barthalion, @nedrichards 
In contrast with a couple of other packaging systems, Flatpak-ing seems designed to facilitate the process rather than a serious of tests to catch out those who aren't members of the inner circle.

Thank you very much for your help and your patience with this ignorant newbie!
